### PR TITLE
move filter before to prevent unneeded processing

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
**Please provide a description of this PR:**

### CODE REVIEW

**Comment by hzxuzhonghu** [comment]
assign @ramaraochavali

File: pilot/pkg/serviceregistry/kube/controller/controller.go [hunk 1]
```diff
@@ -426,6 +426,17 @@ func (c *Controller) deleteService(svc *model.Service) {
 }
 
 func (c *Controller) addOrUpdateService(curr *v1.Service, currConv *model.Service, event model.Event, updateEDSCache bool) {
+	// instance conversion is only required when service is added/updated.
```
[the remark this answers is not in the document]
Comment (hzxuzhonghu) [reply pilot/pkg/serviceregistry/kube/controller/controller.go:429]: We donot append serviceHandler to serviceentry. in SE handler, we do trigger push directly. In WE handler, we need to notify workload handler, which also does not call servcieHandler

[the remark this answers is not in the document]
Comment (hzxuzhonghu) [reply pilot/pkg/serviceregistry/kube/controller/controller.go:429]: >Like the initDiscoveryNamespaceHandlers will re-enqueue the same Service

What do you mean? In namespace handler, we also call `addOrUpdateService`

[the remark this answers is not in the document]
Comment (howardjohn) [reply pilot/pkg/serviceregistry/kube/controller/controller.go:429]: initDiscoveryNamespaceHandlers will just repush every service which now we will skip immediately I think?

[the remark this answers is not in the document]
Comment (hzxuzhonghu) [reply pilot/pkg/serviceregistry/kube/controller/controller.go:429]: NO, in initDiscoveryNamespaceHandlers, we only have add、delete event， but here we only skip on update event

[the remark this answers is not in the document]
Comment (howardjohn) [reply pilot/pkg/serviceregistry/kube/controller/controller.go:429]: Interesting... that is super super subtle.

**Comment by ramaraochavali** [comment]
How significant is the performance improvement here? If it is not much, I would prefer not to complicate this.

**Comment by hzxuzhonghu** [comment]
It is not related to perf, if no fix it will lead to duplicate workload pushed to ztunnel

**Comment by hzxuzhonghu** [comment]
1. This will save all serviceEntry generated service compare
2. no much improvement to k8s service

**Comment by hzxuzhonghu** [comment]
@ramaraochavali ptal

File: pilot/pkg/serviceregistry/kube/controller/serviceimportcache.go [hunk 2]
```diff
@@ -193,6 +193,7 @@ func (ic *serviceImportCacheImpl) onServiceImportEvent(_, obj controllers.Object
 	// a change to the discoverability policy.
 	ic.addOrUpdateService(nil, mcsService, event, true)
 
+	// TODO: do we really need a full push, we should do it in `addOrUpdateService`.
 	if needsFullPush {
 		ic.doFullPush(mcsHost, si.GetNamespace())
```
Comment (hzxuzhonghu) [inline pilot/pkg/serviceregistry/kube/controller/serviceimportcache.go:198]: we may not need this at all, as we do that in `addOrUpdateService`

**Comment by ramaraochavali** [comment]
will take a look tomorrow

**Review by ramaraochavali — APPROVED** [review]
LGT

**Pull request merged**